hack/test_e2e.sh

```diff
@@ -0,0 +1,277 @@
+: ${E2E_DIR:="${0%/*}/temp/e2e"}
+: ${E2E_BIN:="${0%/*}/temp/e2e/bin"}
+
+FORCE_DOCKER_BUILD=false
+OS=$(uname| tr '[:upper:]' '[:lower:]')
+
+
+function line {
+    echo "+============================================================================+"
+}
+
+function docker_build {
+    line
+    echo "   Resolving kpng docker image"
+    line
+    pushd "${0%/*}/.." > /dev/null
+    docker build -t kpng:test -f Dockerfile .
+    popd > /dev/null
+    echo -e "Let's move on.\n"
+}
+
+# Use for CI.
+function store_docker_image {
+    pushd "${0%/*}/.." > /dev/null
+        mkdir _output
+        docker save kpng:test > _output/kpng-image.tar
+    popd > /dev/null
+}
+
+function is_ipv6_available {
+    return cat /sys/module/ipv6/parameters/disable 
+}
+
+function is_ip4_forwarding_available {
+    return sysctl net.ipv4.ip_forward
+}
+
+function setup_kind {
+    line
+    echo "   Resolving kind bin"
+    line
+    if ! kind > /dev/null 2>&1 ; then
+        echo "kind not found, pulling kind binary"
+        curl -L ${E2E_DIR}/kind https://kind.sigs.k8s.io/dl/v0.11.1/kind-linux-amd64 -o ${E2E_DIR}/kind
+        sudo cp ${E2E_DIR}/kind ${E2E_BIN}/kind
+        alias kind=${E2E_BIN}/kind
+    else
+        echo "kind bin found."
+    fi
+    echo -e "Let's move on.\n"
+}
+
+function setup_kubectl {
+    line
+    echo "   Resolving kubectl bin"
+    line
+    if ! kubectl > /dev/null 2>&1 ; then
+        echo "kubectl not found, pulling kind binary"
+        curl -L https://dl.k8s.io/${{ K8S_VERSION }}/bin/${{ OS }}/amd64/kubectl -o ${E2E_DIR}/kubectl
+        sudo cp ${E2E_DIR}/kubectl ${E2E_BIN}/kubectl
+        alias kubectl=${E2E_BIN}/kubectl
+    else
+        echo "kubectl bin found."
+    fi
+    echo -e "Let's move on.\n"
+}
+
+function setup_test {
+    line
+    echo "   Resolving e2e test bins"
+    line
+    if ! [ -f {${E2E_BIN}/ginkgo} ] || ! [ -f {${E2E_BIN}/e2e.test} ] ; then
+        curl -L https://dl.k8s.io/${E2E_K8S_VERSION}/kubernetes-test-${OS}-amd64.tar.gz \
+            -o ${E2E_DIR}/kubernetes-test-${OS}-amd64.tar.gz
+        tar xvzf ${E2E_DIR}/kubernetes-test-${OS}-amd64.tar.gz \
+            --directory ${E2E_DIR} \
+            --strip-components=3 kubernetes/test/bin/ginkgo kubernetes/test/bin/e2e.test
+        rm ${E2E_DIR}/kubernetes-test-${OS}-amd64.tar.gz
+        sudo cp ${E2E_DIR}/ginkgo ${E2E_BIN}/ginkgo
+        sudo cp ${E2E_DIR}/e2e.test ${E2E_BIN}/e2e.test
+    fi
+    echo -e "Let's move on.\n"
+}
+
+function setup_environment {
+    mkdir -p "${E2E_DIR}/bin"
+
+    setup_kind
+    setup_kubectl
+    setup_test 
+
+    sudo chmod +x "${E2E_BIN}"
+}
+
+function create_cluster {
+    line
+    echo "  Building the cluster '${E2E_CLUSTER_NAME}'"
+    line
+
+    # Get rid of any old cluster with the same name.
+    if kind get clusters | grep -q ${E2E_CLUSTER_NAME}; then
+        kind delete cluster --name ${E2E_CLUSTER_NAME}
+    fi
+    # create cluster
+
+    cat <<EOF | kind create cluster \
```

Comment:
can we reuse the `hack/test.sh` command here or centralize this configuration writing somewhere?